Implement viewport stuffs for window #1718

farodin91 opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6875)

<!-- Reviewable:end -->

farodin91 committed:
viewport

farodin91 committed:
[WIP] Implement viewport stuffs for window #1718

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

farodin91 committed:
[WIP] Implement viewport stuffs for window #1718

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

farodin91 commented:
Current Problem i get 
```
thread 'ScriptTask PipelineId(0)' has overflowed its stack
Servo exited with return value -4
```
if ```window.scroll(0,0)``` is excuted

jdm commented:
It should become clear what the problem is if you run `./mach run --debugger  tests/html/viewport.html` and look at the backtrace when the crash occurs.

jdm commented on tests/html/viewport.html:
```diff
@@ -0,0 +1,22 @@
+        console.log("ScrollX: "+ window.scrollX);
+      }
+      function scroll(){
+        window.scroll(0,0);
```
Oh, you're overwriting window.scroll with this new function :P

farodin91 replied:
ups

farodin91 commented:
Minimal support of Scroll working.
How do I get the height/width of the entire window/document?

jdm commented:
We can use getBoundingClientRect on the document's root element for that, as mentioned on IRC.

Also, instead of merging changes from master into your branch, could you [rebase](http://rypress.com/tutorials/git/rebasing) your changes instead? It makes it much easier to follow your changes when reviewing the code.

farodin91 commented:
getBoundingClientRect doesn't work correct for me. Return (0,0)
glutin
* get_position() isn't updating, is my code correct !!!
* get_outer_size() isn't updating and some as get_inner_size()
* resize isn't implemented
* set_position should be easy to implement

farodin91 commented:
I'm working on Linux

jdm commented:
What if you use getBoundingClientRect on the document's body element instead?

farodin91 commented:
Current status the most work. if glutin implement this stuff

jdm commented:
Thanks for doing this! I'll give it a proper review sometime in the next couple days.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6784) made this pull request unmergeable. Please resolve the merge conflicts.

farodin91 committed:
Merge branch 'window' of github.com:farodin91/servo into window

Conflicts:
	components/compositing/compositor.rs
	components/msg/compositor_msg.rs
	components/script/dom/webidls/Window.webidl
	components/script/dom/window.rs
	components/script/script_task.rs
	tests/html/viewport.html

farodin91 committed:
Merging conflict

farodin91 committed:
Merging conflict 2

farodin91 commented:
Will Squash at the end

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

Conflicts:
	components/script/dom/window.rs

farodin91 committed:
last Conflict

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

jdm commented:
Great work! There are just some details that should be cleaned up for completeness and correctness.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 7 of 12 files at r4, 1 of 3 files at r5, 1 of 1 files at r6, 1 of 2 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 34 unresolved discussions, all commit checks successful.

---

<sup>**[components/compositing/compositor.rs, line 431 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5AyJsCHelX0E9MwAR)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/compositor.rs#L431)):</sup>
nit: move the ShutdownState to the next line.

---

<sup>**[components/compositing/compositor_task.rs, line 15 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5B5cuaD0clUWfhI0g)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/compositor_task.rs#L15)):</sup>
nit: move euclid::size comes after euclid::rect alphabetically.

---

<sup>**[components/compositing/compositor_task.rs, line 217 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5NZfavMPEOjfJ2pIW)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/compositor_task.rs#L217)):</sup>
u32

---

<sup>**[components/compositing/compositor_task.rs, line 219 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5CMDeOyT2x-w2Ax1H)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/compositor_task.rs#L219)):</sup>
Document these, please. Also, u32 instead of i32.

---

<sup>**[components/compositing/headless.rs, line 112 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5CSjrKvUjmSsHPtDs)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/headless.rs#L112)):</sup>
This should probably send values over the channel, or any JS that uses the new properties will hang...

---

<sup>**[components/compositing/windowing.rs, line 9 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5BMELrlVNUr47ELlL)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/windowing.rs#L9)):</sup>
nit: euclid::rect comes after euclid::point alphabetically.

---

<sup>**[components/compositing/windowing.rs, line 107 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5BWyb0gap-pUpknUj)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/windowing.rs#L107)):</sup>
Outersize is a pretty web-specific term. Can we be more clear about what values this methods returns?

---

<sup>**[components/compositing/windowing.rs, line 108 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5Ntrdc2ab8_De-awy)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/windowing.rs#L108)):</sup>
u32 instead of i32.

---

<sup>**[components/compositing/windowing.rs, line 110 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5BSlhNThTTfvGkL_E)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/compositing/windowing.rs#L110)):</sup>
u32 instead of i32.

---

<sup>**[components/msg/compositor_msg.rs, line 125 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5NyUv7wOZupgAmHr_)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/msg/compositor_msg.rs#L125)):</sup>
u32

---

<sup>**[components/msg/compositor_msg.rs, line 127 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5Nz0kCTmE7YyBnvMF)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/msg/compositor_msg.rs#L127)):</sup>
u32

---

<sup>**[components/script/dom/webidls/Window.webidl, line 135 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5DGvT2eeibF1OmVcO)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/webidls/Window.webidl#L135)):</sup>
The spec doesn't have a third argument here.

---

<sup>**[components/script/dom/webidls/Window.webidl, line 137 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5DJydlU8gzvWsGGJx)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/webidls/Window.webidl#L137)):</sup>
No third argument in the spec.

---

<sup>**[components/script/dom/webidls/Window.webidl, line 139 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5DLsTguJ1SKma6ru1)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/webidls/Window.webidl#L139)):</sup>
No third argument in the spec.

---

<sup>**[components/script/dom/window.rs, line 11 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5DRSexGfWJNd0TVEV)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L11)):</sup>
nit: we use multiple lines of `use a::b::{c, d}` rather than splitting a single use statement over multiple lines.

---

<sup>**[components/script/dom/window.rs, line 223 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5DiV2ItksrLZHDKPW)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L223)):</sup>
Document this, please. A better name might be `current_viewport`.

---

<sup>**[components/script/dom/window.rs, line 643 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5KOSswKqrt90tLgEy)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L643)):</sup>
Since this method shouldn't take three arguments, we should add a helper method (not in WebIDL) that does and call that from here (and Scroll). We can create default ScrollOptions values with `ScrollOptions::empty()`.

---

<sup>**[components/script/dom/window.rs, line 644 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5EgE9GLqlTVb8IpLe)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L644)):</sup>
Step 3 is missing.

---

<sup>**[components/script/dom/window.rs, line 649 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5FWGNrFnRdag9DsKB)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L649)):</sup>
Let's make this `// Step 5` and `// TODO remove scrollbar width` on separate lines.

---

<sup>**[components/script/dom/window.rs, line 651 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5Fb-_eJizbkKpMCtP)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L651)):</sup>
Same here.

---

<sup>**[components/script/dom/window.rs, line 657 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5GHfzpa81ntj838Wu)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L657)):</sup>
nit: space after ,
Let's also call these x and y, like the spec.

---

<sup>**[components/script/dom/window.rs, line 665 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5IvgNUrn50qbuqsWW)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L665)):</sup>
nit: indent the second line by one space

---

<sup>**[components/script/dom/window.rs, line 678 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5HDJNlYLd_C7XFRKE)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L678)):</sup>
This is just `let document = self.Document();`

---

<sup>**[components/script/dom/window.rs, line 680 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5JsPXl8a69iBzNMqL)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L680)):</sup>
Let's add a `perform_a_scroll` method which implements the steps at https://drafts.csswg.org/cssom-view/#perform-a-scroll (including the call to scroll_fragment_point).

---

<sup>**[components/script/dom/window.rs, line 682 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5JCSBotdkV6N4eqwh)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L682)):</sup>
Let's make scroll_fragment_point (and ScrollFragmentPoint) take a ScrollBehaviour enum that is based on the value that is passed in from JS. We can ignore it in the code that handles ScrollFragmentPoint for now, but it makes it easier to add support in the future.

---

<sup>**[components/script/dom/window.rs, line 718 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5MDWvLaWVlmYvUg6e)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L718)):</sup>
Add `//TODO determine if this operation is allowed`

---

<sup>**[components/script/dom/window.rs, line 725 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5M4KFxxrPpuOIPhwq)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L725)):</sup>
ReszieTo?

---

<sup>**[components/script/dom/window.rs, line 732 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5Ls9gLQro8oEJP2ex)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L732)):</sup>
Add `//TODO determine if this operation is allowed`

---

<sup>**[components/script/dom/window.rs, line 768 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5LFzu4ekIUXlSRa_c)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L768)):</sup>
`self.window_size.get().map(|data| data.device_pixel_ratio).unwrap_or(1.0)`

---

<sup>**[components/script/dom/window.rs, line 998 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5MRdQ59qnnr3TMan5)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L998)):</sup>
Just call unwrap here instead of trying to handle an error.

---

<sup>**[components/script/script_task.rs, line 209 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5MjTfLpBjsPatzR7f)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/script_task.rs#L209)):</sup>
This is unused.

---

<sup>**[ports/cef/window.rs, line 23 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5MpvxUlwzOCYcUzxM)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/ports/cef/window.rs#L23)):</sup>
nit: point and rect are earlier than scale_factor alphabetically.

---

<sup>**[ports/cef/window.rs, line 226 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5N6M03U-_KcZ1IKmi)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/ports/cef/window.rs#L226)):</sup>
We can base this on `self.size()`

---

<sup>**[ports/gonk/src/window.rs, line 798 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5OTIgXIuxTL7L9jlv)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/ports/gonk/src/window.rs#L798)):</sup>
We can base this on self.size()

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

farodin91 committed:
review changes

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

farodin91 committed:
current state

farodin91 commented:
Review status: 2 of 12 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/window.rs, line 644 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5EgE9GLqlTVb8IpLe)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L644)):</sup>
Normalize non-finite values for x and y. 
i don't know, what i have to do?

---

<sup>**[components/script/dom/window.rs, line 682 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5JCSBotdkV6N4eqwh)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L682)):</sup>
Is possible to convert the Bindings enum in other enum

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

jdm commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 10 of 11 files at r10.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/window.rs, line 644 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5EgE9GLqlTVb8IpLe)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L644)):</sup>
According to https://drafts.csswg.org/cssom-view/#normalize-non_finite-values, use http://doc.servo.org/num/traits/trait.Float.html#tymethod.is_finite and set to zero otherwise.

---

<sup>**[components/script/dom/window.rs, line 669 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8I_R5xT-HR8vdJ2bl)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/components/script/dom/window.rs#L669)):</sup>
This doesn't look like it compiles :)

---

<sup>**[components/script/dom/window.rs, line 682 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5JCSBotdkV6N4eqwh)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L682)):</sup>
I'm not sure what you're asking.

---

<sup>**[components/script/dom/window.rs, line 741 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8J02ufIv9XO0Iyqje)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/components/script/dom/window.rs#L741)):</sup>
This can just be `as f64`.

---

<sup>**[components/script/dom/window.rs, line 854 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8J6rPKS9EFvXFhVby)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/components/script/dom/window.rs#L854)):</sup>
nit: newline above this

---

<sup>**[components/script/dom/window.rs, line 888 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8JGG96TJfx6nu_hUi)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/components/script/dom/window.rs#L888)):</sup>
y == ScrollY()

---

<sup>**[ports/cef/window.rs, line 229 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8JdFoeYHIWqgaCnpl)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/ports/cef/window.rs#L229)):</sup>
Add a TODO here to get the real window position.

---

<sup>**[ports/glutin/window.rs, line 522 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8JiqmEKeDykfiVf79)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/ports/glutin/window.rs#L522)):</sup>
I don't think we need casts here any more.

---

<sup>**[ports/glutin/window.rs, line 525 \[r10\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw8JhAjJN3zcLkrKBUy)** ([raw file](https://github.com/servo/servo/blob/225d2ee07d05527d80c4d25a9a1b53a92e6872b7/ports/glutin/window.rs#L525)):</sup>
nit: space after ,

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

farodin91 committed:
most changes to the last review

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

farodin91 commented:
Review status: 9 of 12 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/window.rs, line 650 \[r9\]](https://reviewable.io:443/reviews/servo/servo/6875#-Jw5JCSBotdkV6N4eqwh)** ([raw file](https://github.com/servo/servo/blob/8d2008e0ed7770077e3cfb8182b53803b2e80062/components/script/dom/window.rs#L650)):</sup>
I find a soluation for my problem 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

jdm commented:
Just waiting on changes to scroll_fragment_point/ScrollFragmentPoint now.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 4 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

farodin91 committed:
init support of smooth

jdm commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/msg/compositor_msg.rs, line 143 \[r12\]](https://reviewable.io:443/reviews/servo/servo/6875#-JwHiph3uThViV2ratui)** ([raw file](https://github.com/servo/servo/blob/9a803396f92cfb2de6853f976fcf27bcbb427281/components/msg/compositor_msg.rs#L143)):</sup>
Let's call this `_smooth`

---

<sup>**[components/script/dom/window.rs, line 904 \[r12\]](https://reviewable.io:443/reviews/servo/servo/6875#-JwHi0pzEUc1xy5q20TO)** ([raw file](https://github.com/servo/servo/blob/9a803396f92cfb2de6853f976fcf27bcbb427281/components/script/dom/window.rs#L904)):</sup>
This should take an `Option<&Element>` argument too.

---

<sup>**[components/script/dom/window.rs, line 908 \[r12\]](https://reviewable.io:443/reviews/servo/servo/6875#-JwHiHPwKjwJMqxWA6v7)** ([raw file](https://github.com/servo/servo/blob/9a803396f92cfb2de6853f976fcf27bcbb427281/components/script/dom/window.rs#L908)):</sup>
This should check be true if there's an element, and have a TODO for computing the scroll-behaviour CSS value like the spec says.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

farodin91 committed:
review changes

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7191) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Sorry, I didn't notice that new changes had been made. One last one and we should be good to merge!
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 4 of 4 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

<sup>**[components/script/dom/window.rs, line 908 \[r12\]](https://reviewable.io:443/reviews/servo/servo/6875#-JwHiHPwKjwJMqxWA6v7)** ([raw file](https://github.com/servo/servo/blob/9a803396f92cfb2de6853f976fcf27bcbb427281/components/script/dom/window.rs#L908)):</sup>
Let's have code like this:
```rust
ScrollBehaviour::Auto => {
    element.map(|element| {
        // TODO check computed scroll-behaviour CSS property
        true
    }).unwrap_or(false)
}
```

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

farodin91 committed:
Merging and fix review

Conflicts:
	components/msg/compositor_msg.rs
	components/script/dom/window.rs
	components/script/script_task.rs

farodin91 committed:
fix tidy

farodin91 commented:
I fix a problem with updating of ScrollX and ScrollY, but doesn't if this the correct handling.

farodin91 commented:
r? @jdm

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

farodin91 committed:
fixing borrow_mut

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7265) made this pull request unmergeable. Please resolve the merge conflicts.

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

Conflicts:
	components/compositing/compositor.rs
	components/msg/compositor_msg.rs

jdm commented:
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 9 files at r14, 11 of 11 files at r16.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/compositing/compositor_task.rs, line 69 \[r16\]](https://reviewable.io:443/reviews/servo/servo/6875#-JxNU_-kVp5Asmnoxvz0)** ([raw file](https://github.com/servo/servo/blob/3f3b1762757b9445791d311f8c4fb6c3134b3aea/components/compositing/compositor_task.rs#L69)):</sup>
We should send the smooth value here too.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7401) made this pull request unmergeable. Please resolve the merge conflicts.

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

Conflicts:
	components/script/dom/window.rs

dzbarsky commented on components/script/dom/window.rs:
```diff
@@ -589,8 +595,152 @@ impl<'a> WindowMethods for &'a Window {
+
+    // https://drafts.csswg.org/cssom-view/#dom-window-devicepixelratio
+    fn DevicePixelRatio(self) -> Finite<f64> {
+        let dpr = self.window_size.get()
```
This will return the wrong DPR if we are running with an option specifiying DPR. (See device_pixels_per_screen_px in compositor.rs: http://mxr.mozilla.org/servo/source/components/compositing/compositor.rs#1139)
Not sure how much we care though.

jdm replied:
This value comes from http://mxr.mozilla.org/servo/source/components/compositing/compositor.rs#779 which is derived from device_pixels_per_screen_px. Why is this wrong?

dzbarsky replied:
Aaah I'm having trouble reading. Ignore me and carry on.

jdm commented:
Time to squash! Thanks!
-S-needs-rebase -S-needs-rebase +S-needs-squash 

---

Reviewed 6 of 6 files at r17.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6875)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7417) made this pull request unmergeable. Please resolve the merge conflicts.

farodin91 committed:
Merge remote-tracking branch 'upstream/master' into window

Conflicts:
	components/compositing/headless.rs
	components/script/dom/window.rs
